File: problem/Smallest_Subtree/subtreeWithAllDeepest.js
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var subtreeWithAllDeepest = function(root) {
    if (!root) return null;

    let parent = new Map();
    let queue = [root];
    parent.set(root, null);
    let lastLevel = [];

    while (queue.length > 0) {
        let size = queue.length;
        lastLevel = [...queue];
        for (let i = 0; i < size; i++) {
            let node = queue.shift();
            if (node.left) {
                parent.set(node.left, node);
                queue.push(node.left);
            }
            if (node.right) {
                parent.set(node.right, node);
                queue.push(node.right);
            }
        }
    }

    let deepest = new Set(lastLevel);
    while (deepest.size > 1) {
        let next = new Set();
        for (let node of deepest) {
            next.add(parent.get(node));
        }
        deepest = next;
    }

    return Array.from(deepest)[0];
};

File: problem/Smallest_Subtree/subtreeWithAllDeepest.py
class Solution:
    def subtreeWithAllDeepest(self, root):
        if not root: return None
        
        parent = {root: None}
        queue = [root]
        last_level = []
        
        while queue:
            last_level = queue[:]
            next_queue = []
            for node in queue:
                if node.left:
                    parent[node.left] = node
                    next_queue.append(node.left)
                if node.right:
                    parent[node.right] = node
                    next_queue.append(node.right)
            queue = next_queue
            
        deepest = set(last_level)
        while len(deepest) > 1:
            next_set = set()
            for node in deepest:
                next_set.add(parent[node])
            deepest = next_set
            
        return list(deepest)[0]